### PR TITLE
[bug] - Optimize DiskBufferReader's Seek method to reduce memory usage

### PR DESCRIPTION
**Pull request by ahrav**
This PR introduces an optimization to the `Seek` method of the DiskBufferReader. Previously, when seeking beyond the current read position, the method was inefficient in terms of memory usage, as it required allocating and managing large byte slices to skip unwanted data. This could lead to excessive memory consumption, especially when large skips were involved.

Key Changes:
- Elimination of Large Buffer Allocations: The new implementation removes the need for allocating large byte slices (trashBytes). Instead of reading data into a buffer and then discarding it, we now directly discard the data using io.Discard.
- Use of io.CopyN with io.Discard: To efficiently skip over bytes, io.CopyN is used in conjunction with io.Discard. This method allows us to precisely skip the desired amount of data without actually loading it into memory.

Screenshot 2024-04-16 at 5 56 16 PM

Bechmarks:
Screenshot 2024-04-17 at 12 21 24 PM

### CODE REVIEW

**Review by rosecodym — COMMENTED** [review]
I feel like I must be missing something obvious because it reads to me like you used to throw away data and now you don't, which seems like a pretty big change to not be called out in the PR description. Where have I gone wrong?

**Comment by ahrav** [comment]
> I feel like I must be missing something obvious because it reads to me like you used to throw away data and now you don't, which seems like a pretty big change to not be called out in the PR description. Where have I gone wrong?

Whoops, yea my bad.

File: disk_buffer_reader.go [hunk 1]
```diff
@@ -118,44 +118,53 @@ func (dbr *DiskBufferReader) Reset() error {
...
+		newIndex = dbr.bytesRead + offset
+	default:
+		return 0, fmt.Errorf("invalid whence value")
+	}
+
+	if newIndex < 0 {
+		return 0, fmt.Errorf("can not seek to before start of reader")
+	}
+
+	// If seeking past the bytes read and recording is on, fill the gap by reading the necessary bytes.
```
Comment (rosecodym) [inline disk_buffer_reader.go]: i think this comment is out of date

**Review by rosecodym — COMMENTED** [review]
why aren't you using `io.Discard`? is it just beacuse we'll eventually want to read those bytes?

**Comment by ahrav** [comment]
> why aren't you using `io.Discard`? is it just beacuse we'll eventually want to read those bytes?

That's what I initially thought, but now that I look at it again, i'm not so sure. I _think_ we could use `io.Discard`. I think the re-reads would take place from the tempFile. I'll confirm with Bill.

**Comment by ahrav** [comment]
@bill-rich any reason why we can't use `io.Discard` here with `io.CopyN` vs a Read? i'm guessing it's b/c what Cody mentioned that we need to eventually read those bytes?

**Comment by ahrav** [comment]
> > why aren't you using `io.Discard`? is it just beacuse we'll eventually want to read those bytes?
> 
> That's what I initially thought, but now that I look at it again, i'm not so sure. I _think_ we could use `io.Discard`. I think the re-reads would take place from the tempFile. I'll confirm with Bill.

Well... looks like even using discard does a read from the reader 🤣 .
Screenshot 2024-04-17 at 8 15 01 PM

**Pull request closed without merging**